DEVELOPERS.md

```diff
@@ -0,0 +1,203 @@
+# Development on puppet-agent module
+
+Welcome to the `puppet-agent` module! This is a document which will help
+guide you to getting started on development. The key sections here are:
+
+1. [Iterative code changes](#iterative)
+1. [Important files to understand repo](#files)
+1. [Tools for development](#tools)
+1. [Testing](#testing)
+
+# Iterative code changes<a name="iterative"/>
+
+/*Scott*/
+For iterative development on Ubuntu using Docker, see:
+- docker/ubuntu/Dockerfile
+- docker/bin/run-ubuntu.sh
+
+/*Casey*/
+1. Check out a vmpooler VM.
+2. Manually install PE on it (download from getpe.delivery.puppetlabs.net)
+3. Manually install local copy of the puppet_agent module on it
+4. Request another VM (the agent)
+5. Use the puppet_agent bolt task to install an agent on second VM
+6. Associate the master with an agent
+7. Write a manifest to test what I'm working on
+8. Run `puppet apply` with that manifest
+9. Edit the module on the master, and repeat.
+
+/*Branan*/
+1. Forward your ssh key to do github operations
+2. Ensure editor suppports editing remote files
+3. Checkout a vmpooler machine
+4. Clone the module
+5. Run `puppet apply` to test it as I work
+
+/*Enis*/
+1. Install an "old" PE on a master (typically RHEL).
+2. Declare the pe_repo classes for each agent you're testing upgrades for (e.g. like Windows, Fedora, etc.) using the PE Console.
+3. Install a "new" PE on the same master.
+4. Run puppet agent so that the pe_repo platforms have the new agent package.
+5. Copy local puppetlabs-puppet_agent module over to the /etc/puppetlabs/code/environment/production/modules directory.
+6. Install any module dependencies on the master via puppet module install (which you can get from the metadata.json file).
+7. Grab a VM to use as the agent, either in a local hypervisor or in VCloud.
+8. Test an upgrade scenario by installing the "old" agent by curling the old PE's install.bash script onto the agent VM and running puppet agent -t to link up with the master.
+9. Declare the puppet_agent class in the site manifest (site.pp) declaring the master's agent as the puppet agent version for the agent node.
+10. On the agent node, run `puppet agent -t`.
+11. To check that the upgrade was performed successfully, check the aio_agent_version to see if it matches what's reported on the master.
+
+# Important files to understand repo<a name="files"/>
+
+These are a few key files to look at to understand the basic flow of how the repository works:
+
+- manifests/init.pp -- This is the entrypoint for the module. This file
+  describes the module's parameters, includes other classes(e.g.
+  `prepare` and `install`), and adds some exception cases for a variety
+  of platforms.
+- /*TODO*/
```

Comment:
Which other files do folks find representative of or informative about this repo?